core/arch/aarchxx/mangle.c

```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
+                         opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0,
+                                               -stack_inc, OPSZ_PTR),
+                         opnd_create_reg(DR_REG_XSP),
+                         opnd_is_reg(*arg0) ? *arg0 : opnd_create_reg(DR_REG_XZR),
+                         opnd_create_reg(DR_REG_XSP),
+                         opnd_create_immed_int(-stack_inc, OPSZ_PTR)));
+                } else {
+                    /* sub sp, sp, #(stack_inc) */
+                    PRE(ilist, instr, INSTR_CREATE_sub
+                        (dcontext, opnd_create_reg(DR_REG_XSP),
+                         opnd_create_reg(DR_REG_XSP), OPND_CREATE_INT32(stack_inc)));
+                }
+            } else if (opnd_is_reglike(*arg0)) {
+                if (i + 1 < n && opnd_is_reglike(*arg1)) {
+                    /* stp x(...), x(...), [sp, #(i * XSP_SZ)] */
+                    PRE(ilist, instr, instr_create_1dst_2src
+                        (dcontext, OP_stp,
+                         opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0,
+                                               i * XSP_SZ, OPSZ_16),
+                         opnd_is_reg(*arg0) ? *arg0 : opnd_create_reg(DR_REG_XZR),
+                         opnd_is_reg(*arg1) ? *arg1 : opnd_create_reg(DR_REG_XZR)));
+                } else {
+                    /* str x(...), [sp, #(i * XSP_SZ)] */
+                    PRE(ilist, instr, instr_create_1dst_1src
+                        (dcontext, OP_str,
+                         opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0,
+                                               i * XSP_SZ, OPSZ_PTR),
+                         opnd_is_reg(*arg0) ? *arg0 : opnd_create_reg(DR_REG_XZR)));
+                }
+            } else if (i + 1 < n && opnd_is_reglike(*arg1)) {
+                /* str x(...), [sp, #((i + 1) * XSP_SZ)] */
+                PRE(ilist, instr, instr_create_1dst_1src
+                    (dcontext, OP_str,
+                     opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0,
+                                           (i + 1) * XSP_SZ, OPSZ_PTR),
+                     opnd_is_reg(*arg1) ? *arg1 : opnd_create_reg(DR_REG_XZR)));
+            }
+        }
+#else /* ARM */
+        /* XXX: We could use OP_stm here, but with lots of awkward corner cases. */
+        PRE(ilist, instr, INSTR_CREATE_sub(dcontext, opnd_create_reg(DR_REG_XSP),
+                                           opnd_create_reg(DR_REG_XSP),
+                                           OPND_CREATE_INT32(stack_inc)));
+        for (i = 0; i < n; i++) {
+            opnd_t arg = args[NUM_REGPARM + i];
+            if (opnd_is_reglike(arg)) {
+                /* str r(...), [sp, #(i * 4)] */
+                PRE(ilist, instr, XINST_CREATE_store
+                    (dcontext, opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0,
+                                                     i * 4, OPSZ_PTR), arg));
+            }
```

Comment:
Ditto: confusing to split up where the stack params are stored: why have the reg params in between?